swap/src/bob/swap.rs

```diff
@@ -359,6 +364,31 @@ where
                 )
                 .await
             }
+            BobState::T1Expired(state4) => {
```

Comment:
I haven't looked too deeply into the state of the codebase now, so this may already be accounted for.

There is a difference between T1 expiring (I assume by this you mean the cancel transaction being publishable) before or after Bob has given Alice his encrypted signature on `bitcoin::TxRedeem`:

- If Bob has not given Alice his encrypted signature yet, then Bob only has to make sure that he gets `TxCancel` and `TxRefund` in before Alice can punish him.
- If Bob _has_ given Alice his encrypted signature already, then there is a race between Alice redeeming the bitcoin and Bob publishing `TxCancel`. It's obviously not in Alice's interest to risk this as she can end up revealing `s_a` via the mempool and not getting the bitcoin, but there's still a chance that after you reach `BobState::T1Expired` Alice redeems.

Tl;dr: I think you should be selecting between cancel and redeem futures and if you find the redeem transaction, you should allow Bob to redeem the monero.

Reply:
Thanks @luckysori, that is an interesting edge case!

I will write down the solution in my own words, please let me know if I misunderstood something.

Case: Bob has already sent `encsig` to Alice AND `t1` expires

Bob has to simultanously try to send `tx_cancel` and watch for `tx_redeem`. If `tx_cancel` is confirmed Bob goes for `refund`. If the BTC `redeem` tx sent by Alice is obseved by Bob, Bob has to redeem the Monero.

I will see to add this tomorrow.

Reply:
>  T1 expiring (I assume by this you mean the cancel transaction being publishable) 

Can you clarify why you have to assume that? This is core of the protocol and I am worried when I read that "t1 expired" is not obviously equivalent to "cancel tx can be published". What edge case am I missing?

Reply:
A proper solution to this edge case is properly calculate safe expiries. Something we can do at a later stage.

Reply:
I think what @luckysori wanted to point out is the fact that it can be a race here - even if Bob does not want that he cannot stop it. 
If Bob has sent the `encsig` to Alice but Alice does not redeem until `t1` has passed Bob will try to get the `cancel` tx in. But Alice might race him with the `redeem` transaction (which would make `cancel` invalid).
If we assume that both Alice and Bob act according to our implementation I don't think we will run into this problem at the moment, but eventually we might have to make Bob's side more resilient against this problem, so he still goes on with `redeem` the XMR in such a case and does not get "stuck" in trying to cancel. 

Shall we just create a ticket for this edge-case for now?

Reply:
> > T1 expiring (I assume by this you mean the cancel transaction being publishable)
> 
> Can you clarify why you have to assume that? This is core of the protocol and I am worried when I read that "t1 expired" is not obviously equivalent to "cancel tx can be published". What edge case am I missing?

I only say assume because of the context i.e. the rest of the code in this PR.

Intuitively (to me) `t0` would actually be the cancel transaction timelock and `t1` would be the punish transaction timelock, but I think you are modelling it differently. I just haven't looked at the codebase deeply enough to understand the new model.

Reply:
First time I here of `t0` :) I always heard of `t1` and `t2`.

Reply:
> A proper solution to this edge case is properly calculate safe expiries. Something we can do at a later stage.

Can you explain how safe expiries help here? I think no matter how safe your expiries are, Alice can always publish the redeem transaction (if she has Bob's encrypted signature) to compete with a cancel transaction. It's just that it's generally not in her interest, but the code should eventually handle the race anyway.

Reply:
> First time I here of `t0` :) I always heard of `t1` and `t2`.

Haha, there goes my confusion then.